### PR TITLE
K8SPS-53: Add on-demand backups

### PR DESCRIPTION
**Pull request by egegunes**

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-60/48c3f5e/config.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-60/48c3f5e/init-deploy.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-60/48c3f5e/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-60/48c3f5e/semi-sync.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-60/48c3f5e/service-per-pod.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-60/48c3f5e/sidecars.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-60/48c3f5e/users.log) |

commit: https://github.com/percona/percona-server-mysql-operator/pull/60/commits/48c3f5e4bf49ab9e9842aa537ea38f7e2f601790
image: `perconalab/percona-server-mysql-operator:PR-60-48c3f5e`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-60/acd5f49/config.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-60/acd5f49/init-deploy.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-60/acd5f49/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-60/acd5f49/semi-sync.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-60/acd5f49/service-per-pod.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-60/acd5f49/sidecars.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-60/acd5f49/users.log) |

commit: https://github.com/percona/percona-server-mysql-operator/pull/60/commits/acd5f4927d5f08c4864ebfb8f618a3541fa6e79a
image: `perconalab/percona-server-mysql-operator:PR-60-acd5f49`

**Review by pooknull — DISMISSED** [review]

File: api/v1alpha1/perconaservermysql_types.go [hunk 1]
```diff
@@ -124,6 +125,30 @@ type PMMSpec struct {
...
+type BackupStorageType string
+
+const (
+	BackupStorageFilesystem BackupStorageType = "filesystem"
+	BackupStorageS3         BackupStorageType = "s3"
+	BackupStorageGCS        BackupStorageType = "gcs"
+	BackupStorageAzure      BackupStorageType = "azure"
+)
+
+type BackupStorageSpec struct {
```
Comment (hors) [inline api/v1alpha1/perconaservermysql_types.go:147]: please see https://github.com/percona/percona-xtradb-cluster-operator/blob/main/pkg/apis/pxc/v1/pxc_types.go#L433-L449 and all needed features like Annotations, Labels and so on

File: build/ps-entrypoint.sh [hunk 2]
```diff
@@ -288,10 +294,10 @@ if [ "$1" = 'mysqld' -a -z "$wantHelp" ]; then
 			CREATE USER 'operator'@'${MYSQL_ROOT_HOST}' IDENTIFIED BY '${OPERATOR_ADMIN_PASSWORD}' ;
 			GRANT ALL ON *.* TO 'operator'@'${MYSQL_ROOT_HOST}' WITH GRANT OPTION ;
 
-			CREATE USER 'xtrabackup'@'localhost' IDENTIFIED BY '${XTRABACKUP_PASSWORD}';
-			GRANT SYSTEM_USER, BACKUP_ADMIN, PROCESS, RELOAD, LOCK TABLES, REPLICATION CLIENT ON *.* TO 'xtrabackup'@'localhost';
-			GRANT SELECT ON performance_schema.log_status TO 'xtrabackup'@'localhost';
-			GRANT SELECT ON performance_schema.keyring_component_status TO 'xtrabackup'@'localhost';
+			CREATE USER 'xtrabackup'@'%' IDENTIFIED BY '${XTRABACKUP_PASSWORD}';
```
Comment (hors) [inline build/ps-entrypoint.sh:297]: I am not sure that we need to have separate user. Maybe we can use existing one

File: deploy/crd.yaml [hunk 3]
```diff
@@ -16,7 +16,14 @@ spec:
     singular: perconaservermysqlbackup
   scope: Namespaced
   versions:
-  - name: v1alpha1
```
Comment (hors) [inline deploy/crd.yaml:19]: please add to shortNames `ps-backups`

File: deploy/cr.yaml [hunk 4]
```diff
@@ -97,3 +97,16 @@ spec:
 
     serverHost: monitoring-service
     serverUser: admin
+
+  backup:
+    image: perconalab/percona-server-mysql-operator:xtrabackup
```
Comment (hors) [inline deploy/cr.yaml:102]: I think we need to use official docker image `percona/percona-xtrabackup` .

**Review by hors — CHANGES_REQUESTED** [review]

File: build/ps-entrypoint.sh (hunk 2 above)
Comment (egegunes) [reply build/ps-entrypoint.sh:297]: By existing one you mean `root` or `operator`? I believe we should have a `xtrabackup` user as in PXC

File: deploy/crd.yaml (hunk 3 above)
Comment (egegunes) [reply deploy/crd.yaml:19]: we already have this as a short name

File: deploy/cr.yaml (hunk 4 above)
Comment (egegunes) [reply deploy/cr.yaml:102]: We'll be able to use official image after https://github.com/percona/percona-docker/pull/599 merged.

For restores I'll see what we can do but it's good to go for backups.

File: api/v1alpha1/perconaservermysql_types.go (hunk 1 above)
Comment (egegunes) [reply api/v1alpha1/perconaservermysql_types.go:147]: https://github.com/percona/percona-server-mysql-operator/pull/60/commits/1789ad956beb6221d12ee16008737653097ed68f

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-60/1789ad9/config.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-60/1789ad9/init-deploy.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-60/1789ad9/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-60/1789ad9/semi-sync.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-60/1789ad9/service-per-pod.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-60/1789ad9/sidecars.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-60/1789ad9/users.log) |

commit: https://github.com/percona/percona-server-mysql-operator/pull/60/commits/1789ad956beb6221d12ee16008737653097ed68f
image: `perconalab/percona-server-mysql-operator:PR-60-1789ad9`

**Review by tplavcic — DISMISSED** [review]
Fix shellcheck comments.

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-60/e474b57/config.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-60/e474b57/init-deploy.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-60/e474b57/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-60/e474b57/semi-sync.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-60/e474b57/service-per-pod.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-60/e474b57/sidecars.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-60/e474b57/users.log) |

commit: https://github.com/percona/percona-server-mysql-operator/pull/60/commits/e474b57d70ba739a43726d1fdd4fb5a87e7dfc49
image: `perconalab/percona-server-mysql-operator:PR-60-e474b57`

**Pull request closed without merging**